openedx/core/djangoapps/content/learning_sequences/api/outlines.py

```diff
@@ -0,0 +1,323 @@
+    return user_course_outline
+
+
+def get_user_course_outline_details(course_key: CourseKey,
+                                    user: User,
+                                    at_time: datetime) -> UserCourseOutlineDetailsData:
+    """
+    Get an outline with supplementary data like scheduling information.
+
+    See the definition of UserCourseOutlineDetailsData for details about the
+    data returned.
+    """
+    user_course_outline, processors = _get_user_course_outline_and_processors(
+        course_key, user, at_time
+    )
+    schedule_processor = processors['schedule']
+
+    return UserCourseOutlineDetailsData(
+        outline=user_course_outline,
+        schedule=schedule_processor.schedule_data(user_course_outline)
+    )
+
+
+def _get_user_course_outline_and_processors(course_key: CourseKey,
+                                            user: User,
+                                            at_time: datetime):
+    full_course_outline = get_course_outline(course_key)
+    user_can_see_all_content = can_see_all_content(user, course_key)
+
+    # These are processors that alter which sequences are visible to students.
+    # For instance, certain sequences that are intentionally hidden or not yet
+    # released. These do not need to be run for staff users. This is where we
+    # would add in pluggability for OutlineProcessors down the road.
+    processor_classes = [
+        ('schedule', ScheduleOutlineProcessor),
+        ('visibility', VisibilityOutlineProcessor),
+        # Future:
+        # ('content_gating', ContentGatingOutlineProcessor),
+        # ('milestones', MilestonesOutlineProcessor),
+        # ('user_partitions', UserPartitionsOutlineProcessor),
+    ]
+
+    # Run each OutlineProcessor in order to figure out what items we have to
+    # remove from the CourseOutline.
+    processors = dict()
+    usage_keys_to_remove = set()
+    inaccessible_sequences = set()
+    for name, processor_cls in processor_classes:
+        # Future optimization: This should be parallelizable (don't rely on a
+        # particular ordering).
+        processor = processor_cls(course_key, user, at_time)
+        processors[name] = processor
+        processor.load_data()
+        if not user_can_see_all_content:
+            # function_trace lets us see how expensive each processor is being.
+            with function_trace('processor:{}'.format(name)):
+                processor_usage_keys_removed = processor.usage_keys_to_remove(full_course_outline)
+                processor_inaccessible_sequences = processor.inaccessible_sequences(full_course_outline)
+                usage_keys_to_remove |= processor_usage_keys_removed
+                inaccessible_sequences |= processor_inaccessible_sequences
+
+    # Open question: Does it make sense to remove a Section if it has no Sequences in it?
+    trimmed_course_outline = full_course_outline.remove(usage_keys_to_remove)
```

Comment:
Thinking out loud, would we want to trim the outline after each iteration of the loop?
That way if one processor removes keys, the subsequent processors don't even see the (already-removed) items. I'm thinking of a whole section being marked as `staff-only`; maybe we can aggressively prune the tree? :shrug: 

Probably not a huge deal; I guess it depends on how much overhead is incurred by the `remove` call...